### ISSUE TITLE
Docstrings needs to be fixed in twisted.words.protocols.irc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6494]
|<img alt="Upasana's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Upasana reported|
|-|-|
|Trac ID|trac#6494|
|Type|enhancement|
|Created|2013-04-28 17:59:41Z|
|Branch|https://github.com/twisted/twisted/tree/irc-docstrings-6494-3|

```
parsemsg: The opening/closing of docstring should be on a line by themselves
IRCClient.ctcpQuery_DCC: The opening/closing of docstring should be on a line by themselves
IRCClient.dccDoSend: The opening/closing of docstring should be on a line by themselves
IRCClient.dccDoResume: The opening/closing of docstring should be on a line by themselves
IRCClient.dccDoAcceptResume: The opening/closing of docstring should be on a line by themselves
IRCClient.ctcpUnknownReply: The opening/closing of docstring should be on a line by themselves
IRCClient.badMessage: The opening/closing of docstring should be on a line by themselves
IRCClient.quirkyMessage: The opening/closing of docstring should be on a line by themselves
IRCClient.handleCommand: The opening/closing of docstring should be on a line by themselves
DccFileReceiveBasic: The opening/closing of docstring should be on a line by themselves
DccFileReceiveBasic.dataReceived: The opening/closing of docstring should be on a line by themselves
DccSendProtocol: The opening/closing of docstring should be on a line by themselves
fileSize: The opening/closing of docstring should be on a line by themselves
DccChat: The opening/closing of docstring should be on a line by themselves
DccChat.__init__: The opening/closing of docstring should be on a line by themselves
dccDescribe: The opening/closing of docstring should be on a line by themselves
DccFileReceive: The opening/closing of docstring should be on a line by themselves
DccFileReceive.set_directory: The opening/closing of docstring should be on a line by themselves
DccFileReceive.set_filename: The opening/closing of docstring should be on a line by themselves
DccFileReceive.set_overwrite: The opening/closing of docstring should be on a line by themselves
DccFileReceive.connectionLost: The opening/closing of docstring should be on a line by themselves

```

Attachments:

* [bug_6494.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afd/afdbd553a7341d5bbf2786985885ff5b760c217e/110e12c09234074b2457cd1bcb865fb3c4db7503.patch) (7567 bytes) - added by Upasana on 2013-04-28 18:06:09Z - 
* [bug_6494_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afd/afdbd553a7341d5bbf2786985885ff5b760c217e/c5adbabf0195e1402f20ba8790ca216fde61d6a1.patch) (14043 bytes) - added by eeshangarg on 2014-08-16 16:16:38Z - 
* [bug_6494_3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afd/afdbd553a7341d5bbf2786985885ff5b760c217e/f7e900014b8ed8330ee1c1903aa75bb92e3f702f.patch) (13854 bytes) - added by eeshangarg on 2014-08-18 17:03:23Z - 
* [bug_6494_4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afd/afdbd553a7341d5bbf2786985885ff5b760c217e/18087c647cb3ff194b3d6a4a35adddaf826af5d6.patch) (14065 bytes) - added by eeshangarg on 2015-01-31 14:14:39Z - 
* [bug_6494_5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afd/afdbd553a7341d5bbf2786985885ff5b760c217e/58fe87bb9a4e209b2bbb4a3f46d3c061bc7af5ae.patch) (1699 bytes) - added by eeshangarg on 2015-02-05 16:56:02Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6494 6494
type__enhancement enhancement
reporter__Upasana Upasana
priority__normal normal
milestone__ 
branch__branches_irc_docstrings_6494_3 branches/irc-docstrings-6494-3
branch_author__adiroiban__glyph adiroiban, glyph
status__closed closed
resolution__fixed fixed
component__words words
keywords__easy_documentation easy documentation
time__1367171981000000 1367171981000000
changetime__1423282113539618 1423282113539618
version__None None
owner__glyph glyph
cc__ralphm cc__thijs cc__eeshangarg cc__lvh
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Upasana's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Upasana removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra set owner to Upasana|
|-|-|

I believe this ticket is about making the docstrings conform to the Twisted coding standard. It would make sense if you could add documentation for the various functions you touched, eg.:

```python
def parsemsg(s):
    """
    Breaks a message from an IRC server into its prefix, command, and arguments.

    @param s: [description here]
    @type s: [type here, eg. L{str}]
    @rtype: [type here, eg. L{tuple}]
    @return: [description here]
    """
```

I expect `twistedchecker` will start complaining about those missing params documentation as soon as we apply this patch, so fixing it directly seems appropriate. If you don't feel up for that task, open a new ticket (with the `easy` keyword) so you or somebody else can pick up this task later). Thanks.

**Comment by twisted-trac** [comment]
|<img alt="Upasana's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Upasana commented|
|-|-|

Alright, I'll add missing docstrings. Thanks for review.

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)|@eeshangarg set owner to @eeshangarg<br>@eeshangarg set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_7"></a>|@eeshangarg set owner to @glyph<br>@eeshangarg set status to `new`|
|-|-|

Hi!

I added params documentation to all of the functions specified in the ticket description. I am completely new to networking and Twisted and to the world of programming in general. I would love to work on this further if needed because I really want to close this ticket and keep contributing to Twisted. I learned a lot while working on it. It was a lot of fun!

The patch file I would like to get reviewed is 'bug_6494_2.patch'.

I couldn't assign the ticket to the blank entry as it says in the docs(I couldn't find the blank entry anywhere). I asked on #twisted-dev about this and was told that this is due to a trac upgrade and was advised to assign the ticket to glyph for the time being. 

Thanks & regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [42995]) Branching to irc-docstrings-6494.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @eeshangarg|
|-|-|

Thanks for your contribution, Eeshan.

I've applied the patch to a branch and kicked off the buildbots, and unfortunately twistedchecker isn't _quite_ happy with it - very close though! [You can see the new errors here.](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/2191/steps/run-twistedchecker/logs/new%20twistedchecker%20errors)

Some additional points:

  1. Although this is inconsistent throughout the code because the policy changed at one point, the current style recommendation for referring to built-in types such as `str` and `tuple` is `L{str}` and `L{tuple}` rather than `C{str}` and C{tuple}`.  This will generate documentation cross-reference links.
  2. Since we are looking forward to a python-3 future (probably in 30 years or so, but still), using `str` to describe the type of an argument is somewhat ambiguous.  We typically use `L{unicode}` and `L{bytes}` to be clear about which one we mean, specifically (although `unicode` isn't actually a defined name in 3.x, it is at least unambiguous).  The only time we would use `str` is to describe the type of things like Python identifiers and the value of `__doc__`, which are bytes on 2.x and strs on 3.x, in which case we'd say "native `str`" to be clear that we mean "we are intentionally being ambiguous about the type because it differs based on your python version" rather than just "`str`" which means "we are being vague about the type because this docstring was written before python 3.x".
  3. As long as you're in here, that "XXX" in the docstring for `ctcpUnknownReply` should be removed and placed in a comment - or, better yet, a ticket, since it's a to-do and not really information about the code itself.  To-dos go in the ticket tracker, not the documentation.
  4. The `excValue` parameter to `badMessage` should have a `@type` as well.  Also, the second argument to `raise` is not "usually" a class instance - for many versions of Python now, it has been required to be an instance of a `BaseException` subclass.  This is what the `@type` should indicate.
  5. The `reason` argument to `connectionLost` is a `Failure` instance.  Why did you think it was a `str`?  Is there some other documentation that needs fixing?

I look forward to re-reviewing this change.  Please attach another patch that addresses the above issues when you have time.

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_11"></a>|@eeshangarg commented|
|-|-|

Hello!

Thanks for reviewing my patch, glyph. It's an honor to have my very first patch to Twisted reviewed by you. I have some specific questions regarding the changes that I need to make. Please allow me to talk about what I understand before I talk about what I am confused about.

I looked at the BuildBot errors, and I will fix the errors in my next patch and I will be more careful and try to avoid my silly mistakes in the future.

I will also need to change all references to built-in types in my patch to use the L{type} syntax.

I will remove the 'XXX' in the docstring for ctcpUnknownReply and file a separate ticket about it.

I will also give the excValue parameter to badMessage a type of L{BaseException}. My mistake here was probably due to my lack of knowledge and I need to learn more and work harder.

I will change the reason argument to connectionLost to have a type of L{Failure}. Me thinking it was an str is also my fault and is downright ignorance on my part. I'll try and avoid such mistakes in the future.

What I am having trouble understanding is how can I determine whether the type of an argument should be a L{bytes} or L{unicode}?. When working on this patch, I read the tests for each function/method I had to write a docstring for and tried to understand what type each argument should have. I would really appreciate it if you could give me some pointer as to how I can determine the type of an argument.

I learned a lot from your advice, sir and I want to keep contributing to Twisted. :)

Thanks & regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_12"></a>|@lvh commented|
|-|-|

Well, unfortunately bytes vs unicode is a tricky problem :)

It boils down to this: think about what a particular object *means*. Is it, semantically, some text? If so, unicode; otherwise, bytes.

Of course, this gets tricky fast. Is an URL bytes, or text? (There's, IIUC, one correct answer.) How about an IRI? How about filesystem paths? Those are bytes on Unix but Unicode on Windows, and I'm not even quite sure on OS X (mostly bytes but with some uppercase/lowercase semantics that seem like they're from text-only?)

This is *particularly* bad for old protocols like FTP and IRC, because they hail from simpler times when everything was probably just ASCII but it's okay; pretend it's bytes and shovel it around. So, even though semantically that text message I am sending to you in a PRIVMSG over IRC is clearly *text* (it's English! you can read it!), it may well be bytes, to IRC.

In IRC, the answer is almost always bytes. A quick grep of the patch suggests that you might be done by simply replacing all the L{str}s with L{bytes}.

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_13"></a>|@eeshangarg commented|
|-|-|

Hi!

I just submitted a new patch. It's called 'bug_6494_3.patch'.

Thanks to lvh for his guidance and mentorship!

Thanks & regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)|@eeshangarg set owner to @glyph|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_16"></a>|@eeshangarg commented|
|-|-|

Hi!

I just filed a ticket regarding the 'XXX' in the docstring for IRCClient.ctcpUnknownReply.

Here is the link to the ticket: [#7560](https://github.com/twisted/twisted/issues/7560)

Please let me know if anything else needs to be done. :)

Thanks & regards,
Eeshan

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph removed owner|
|-|-|

Un-assigning all review tickets assigned to me.  Unless there's a specific reason _I_ need to review something, let's leave them unassigned :-).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_18"></a>|@adiroiban set owner to @adiroiban|
|-|-|

I will try to resolve the conflicts form latest patch and update the branch to see if all is ok...and continue the review

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_19"></a>|@adiroiban commented|
|-|-|

(In [43733]) Branching to irc-docstrings-6494-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_20"></a>|@adiroiban set owner to @eeshangarg|
|-|-|


Thanks for the work on this and sorry for the delay.

I have apply latest patch on a new brand and trigger the tests.

Looks good. Test all green.

Regarding the removal of XXX comment. Not sure if is ok to completly remove it.


```
-        XXX: If the client makes arbitrary CTCP queries,
-        this method should probably show the responses to
-        them instead of treating them as anomolies.
```


I prefer to keep a comment like the one below.
In this way, when someone stumble
upon this problem it is much easier to locate the dedicated ticket and avoid
creating a duplicated ticket.

```
# FIXME:7560:
# Add code for handling arbitrary queries and not treat them as
# anomalies.
```

But this is minor and I think that we can merge it as it is now.

----------

Regarding

```
+    @rtype: L{tuple}
+    @return: A 3-tuple of (prefix, command, args) where args is
+        of type L{list}.
```

Maybe we should put rtype after return and have a simple docstring,
as rtype already has a link to L{tuple}.


```
    @return: A tuple of (prefix, command, args).
    @rtype: L{tuple}
```

Also this is minor and we can merge it.

------------

If you are ok with the changes you can either submit a new patch or I can apply
them and merge it.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_21"></a>|@eeshangarg set owner to @adiroiban|
|-|-|

Hello Mr. adiroiban!

I just added a new patch (bug_6494_4.patch) with all of the changes that you advised me to make. The new patch also includes all of the changes that came with the previous patch, but I resolved all of the conflicts this time. I also included a twisted/words/topfiles/6494.misc file, but whenever I apply the patch to trunk (using patch -p0), the .misc file does not get created due to tool problems.

Please let me know if anything else needs to be done or if something could be better! I would love to work on it! :-)

Warm regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_22"></a>|@adiroiban commented|
|-|-|

(In [43797]) Branching to irc-docstrings-6494-3.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_23"></a>|@adiroiban set owner to @eeshangarg|
|-|-|

Thanks for the patch. I will handle the misc file ... i can see it in the github diff :)

For https://github.com/twisted/twisted/compare/trunk...irc-docstrings-6494-3#diff-17b06a87bc9a7bc8bbac5bd6df97285eR2625

```@ivar bytesReceived: See L{bytesReceived} parameter of L{__init__} ```

I am not sure how you ended up with this. bytesReceived is not an argument of __init__. You should document it in the class docstring.


---------

https://github.com/twisted/twisted/compare/trunk...irc-docstrings-6494-3#diff-17b06a87bc9a7bc8bbac5bd6df97285eR2630

``` def __init__(self, resumeOffset=0): ```

resumeOffset is not documented

-----------

https://github.com/twisted/twisted/compare/trunk...irc-docstrings-6494-3#diff-17b06a87bc9a7bc8bbac5bd6df97285eR2639

``` def dataReceived(self, data): ```

data is not documented

--------

https://github.com/twisted/twisted/compare/trunk...irc-docstrings-6494-3#diff-17b06a87bc9a7bc8bbac5bd6df97285eR2342
u A 3-list of [fileName, address, port] }}}

Does it needs to be a list? I think that a tuple should also work.. or any sequence

Now we are documenting existing code..so it might require a list.

Just a comment for the future, when possible we should document the most generic data structure.

For example if a method uses a list just to iterate over its element, the input arguments should be documented as an iterator, rather than a specific list.
ho-------

https://github.com/twisted/twisted/compare/trunk...irc-docstrings-6494-3#diff-17b06a87bc9a7bc8bbac5bd6df97285eR2486

``` @param line: The indecipherable message. ```

Why this comes without a type?

----------

Not sure why buildbot was happy with the latest patch.

Please address the comments and send a new patch. If possible, please only submit a patch on top of current patch.
Let me know if you need help with that.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_24"></a>|@eeshangarg set owner to @adiroiban|
|-|-|

Hello Mr. adiroiban!

I just submitted the new patch (bug_6494_5.patch) that we had a discussion about on IRC. Thanks a lot for your guidance! Please let me know if anything else needs to be done, I would love to work on this further. :-)

Warm regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_25"></a>|@adiroiban removed owner|
|-|-|

Thanks for the patch.

For 
```
@param resumeOffset: An integer representing the amount of bytes 
   from where the transfer of data should be resumed. 
```

reading the code, I can see that this is not what the code is doing.
When resumeOffset is not 0, it will just open the file in append mode... this is an naive resume.

If the file has 15bytes, you have already sent 10bytes and then ask a resume at 8bytes following another 7byte, the file will be opened in append resulting a final file of 17bytes

now, maybe this is a bug, but without proper documentation is hard to say if this is a bug or not.

------

Regarding ``` See: L{protocol.Protocol.dataReceived} ``` we will need to check with the other developers about what is the right format ... but I think that your patch is good.

------

I am leaving this for review so that another developer can do the final checks.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_26"></a>|@adiroiban commented|
|-|-|

I have also launched a new buildbot run for the latest patch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph set owner to @glyph|
|-|-|

The builds basically look good, and while I can think of more stuff to document here, this is a big improvement to understanding the existing code, so I will land it.

There are a few places where wrapping / indentation is a little bit off (indentation by a non-multiple of 4 within a docstring, for example) so I will fix that before landing.  But this is not even reported by the tools, so I could not expect you to get it right :).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph set status to `closed`|
|-|-|

(In [43873]) Merge irc-docstrings-6494-3: Improve the docstrings in twisted.words.irc.

Author: eeshangarg, Upasana

Reviewer: adiroiban, glyph, thijs

Fixes: [#6494](https://github.com/twisted/twisted/issues/6494)

This adds documentation for many undocumented parameters and brings the
docstrings much more into line with the coding standard.
